### ISSUE TITLE
PT_REGS_PARM macro does not seem to return function arguments in attach_uprobe

### ISSUE THREAD

**Issue by hsane2001** [iovisor/bcc#3056]
Trying to instrument memcpy from standard C library using an uprobe and extract the destination and size arguments. 
memcpy(dest, src, size) - I assume PT_REGS_PARM1 hods the first arg (dest) and PT_REGS_PARM3 holds the size.

char buf[256] = {};
u64 size;
bpf_probe_read(&buf, sizeof(buf), (void *)PT_REGS_PARM1(ctx));
bpf_probe_read(&size, sizeof(size), (void *)PT_REGS_PARM3(ctx));
bpf_trace_printk("buf: %lx, size: %ld\\n",buf, size);

Output of a simple memcpy program:
PID 71137: Memcopying 268435456 bytes from 0x7fbb228d9010 to 0x7fbb208d8010

Kernel trace output:
<...>-73117 [050] .... 3892834.828501: 0: buf: ffffb6c70d31fba0, size: 0

Am I using the trace flags incorrectly, since the trace output seems off.

Thanks in advance,
Harshad

**Comment by yonghong-song** [comment]
The following is not correct
```
 bpf_probe_read(&size, sizeof(size), (void *)PT_REGS_PARM3(ctx));
```
Just do
```
size = PT_REGS_PARM3(ctx);
```

param3 holds the value of `size`, not the address of `size`.

**Comment by hsane2001** [comment]
You're right, however even that comes up with size as always the same number:
<...>-86425 [035] .... 3973364.289879: 0: buf: , size: **1099511627778**

Even if I change the size of memcpy, the trace displays the PT_REGS_PARM3 to always come up as **1099511627778**
Also, why does the source buffer string (PT_REGS_PARM2) come back as empty?

**Comment by yonghong-song** [comment]
Maybe you can make a self-contained test case so people can help take a look?

**Comment by hsane2001** [comment]
Hi,

I simply added the following lines in the "trace_func_entry" of funclatency.py from bcc/tools:

    u32 size;
    size = PT_REGS_PARM3(ctx);
    bpf_trace_printk("size=%d\\n",size);

Then I invoke funclatency.py as follows:  sudo python3 funclatency.py 'c:memcpy' and run my memcpy program, while looking at the output trace.

My memcpy program:
`

    int main(void) {
        char src[64] ="Testing memcpy program";
        char dst[64] = {0};

        printf("Moving %zu Bytes from %p to %p\n", sizeof(src), (void *)&src, (void *)&dst);
        memcpy(dst,src,sizeof(src));

        return 0;
    }
`

No matter what size of memcpy I do, I always get size = 2
memcpy-19894 [017] .... 91470.124731: 0: size=2
memcpy-19895 [065] .... 91471.397751: 0: size=2

**Comment by yonghong-song** [comment]
Could you double check your `main` function `memcpy` exist in the object file? Since the size only 64 bytes, the compiler may optimize it away. The command line is to tracing c library memcpy, there is a possibility that memcpy captured in trace_pipe may be from somewhere else.

**Comment by hsane2001** [comment]
I am tracing memcpy from C itself. Checked the object file and it has memcpy call. Also ran several workloads that do memcpy and the trace still shows the size either = 0 or 2. 
I also tried 1. many memcpy sizes, 2. Turning off gcc optimizations - either way size comes out the same 0 or 2.

**Comment by hsane2001** [comment]
Any ideas?

**Comment by yonghong-song** [comment]
Okay, I did an experiment based on your above description and actually I got nothing from the trace_pipe file.
Further I checked symbol memcpy of libc on my system, I found
```
000000000008a950 i __GI_memcpy                                                                              
000000000008a950 i memcpy
0000000000162320 t __memcpy_avx512_no_vzeroupper
0000000000162d20 t __memcpy_avx512_unaligned
0000000000162d90 t __memcpy_avx512_unaligned_erms
000000000015dfa0 t __memcpy_avx_unaligned
000000000015e010 t __memcpy_avx_unaligned_erms
000000000010ace0 i __memcpy_chk
0000000000162310 t __memcpy_chk_avx512_no_vzeroupper
0000000000162d10 t __memcpy_chk_avx512_unaligned
0000000000162d80 t __memcpy_chk_avx512_unaligned_erms
000000000015df90 t __memcpy_chk_avx_unaligned
000000000015e000 t __memcpy_chk_avx_unaligned_erms
00000000000a3840 t __memcpy_chk_erms
000000000010ace0 t __memcpy_chk_ifunc
00000000000a37d0 t __memcpy_chk_sse2_unaligned
00000000000a38b0 t __memcpy_chk_sse2_unaligned_erms
000000000013b610 t __memcpy_chk_ssse3
0000000000140ba0 t __memcpy_chk_ssse3_back
00000000000a3850 t __memcpy_erms
000000000008a950 i memcpy@@GLIBC_2.14
00000000000a37e0 T memcpy@GLIBC_2.2.5
00000000000a37e0 t __memcpy_sse2_unaligned
00000000000a38c0 t __memcpy_sse2_unaligned_erms
000000000013b620 t __memcpy_ssse3
0000000000140bb0 t __memcpy_ssse3_back
000000000008a950 i __new_memcpy
000000000008a950 t __new_memcpy_ifunc
00000000000a4b00 W wmemcpy
00000000000a4b00 t __wmemcpy
000000000010c580 T __wmemcpy_chk
```

The `memcpy` is a indirect function. Normal tracing like the above won't work.
strlen is another example of indirect function. 
bcc/examples/tracing/strlen_hist_ifunc.py contains an example to trace libc indirect function strlen.
Maybe you can take a look and modify the example for your purpose?

**Comment by hsane2001** [comment]
@yonghong-song - I am surprised you got nothing from the trace_pipe? I dont think memcpy is a indirect function. I tried running "bcc/examples/tracing/strlen_hist_ifunc.py" and it tells me that its not an indirect function. Here are 2 more reasons I dont think so:
1. "nm" on my binary only gives one reference to memcpy.
0000000000001189 T main
                 U memcpy@@GLIBC_2.14
                 U printf@@GLIBC_2.2.5
2. When I run bcc/tools/funclatency, I am able to instrument "c:memcpy" and collect duration, count, distribution etc. correctly with the same memcpy example I stated above.
python3 funclatency.py c:memcpy
Tracing 2 functions for "c:memcpy"... Hit Ctrl-C to end.
Function = b'[unknown]' [5011]
     nsecs               : count     distribution
         0 -> 1          : 0        |                                        |
         2 -> 3          : 0        |                                        |
         4 -> 7          : 0        |                                        |
         8 -> 15         : 0        |                                        |
        16 -> 31         : 0        |                                        |
        32 -> 63         : 0        |                                        |
        64 -> 127        : 0        |                                        |
       128 -> 255        : 0        |                                        |
       256 -> 511        : 0        |                                        |
       512 -> 1023       : 0        |                                        |
      1024 -> 2047       : 0        |                                        |
      2048 -> 4095       : 0        |                                        |
      4096 -> 8191       : 1        |****************************************|
      8192 -> 16383      : 0        |                                        |
     16384 -> 32767      : 1        |****************************************|

Its only failing when I try to derive the PARMS in memcpy. Seems to me like extracting the function arguments is not possible for some reason.

**Comment by yonghong-song** [comment]
The glibc version difference. Mine is libc-2.28.so and yours should be 2.14.

Since my libc won't work, I changed prog to below to chase function `my_memcpy`.
```
$ cat t.c
#include <stdio.h>
#include <string.h>

void __attribute__((noinline)) my_memcpy(char *dst, char *src, int len) {
  memcpy(dst, src, len);
}


int main(void) {
  char src[64] ="Testing memcpy program";
  char dst[64] = {0};

  printf("Moving %zu Bytes from %p to %p\n", sizeof(src), (void *)&src, (void *)&dst);
  my_memcpy(dst,src,sizeof(src));

 return 0;
}
$
```

Command line:
```
./funclatency.py /home/yhs/work/test/bcc/3056/a.out:my_memcpy
```
The trace_pipe output:
```
$ sudo cat /sys/kernel/debug/tracing/trace_pipe
           <...>-462739 [001] d... 323072.424418: bpf_trace_printk: size=64

           <...>-462880 [021] d... 323075.421240: bpf_trace_printk: size=64

^C
```

it must still related to your libc, maybe you do assembly code of memcpy to find what exactly the arguments look like. Maybe glibc did some transformation in the middle.

**Comment by hsane2001** [comment]
@yonghong-song  You are right, I was able to get the src,dest pointer addresses as well as size correctly with an indirect reference to memcpy like you have above. I am still trying to find how/which "memcpy in glibc " is my function calling and how to instrument it, since that is the end goal.

**Comment by sharksforarms** [comment]
I hit this issue while trying to figure out why the strlen_count.py example was not working, I ended up with the same solution (wrapper function). Any ideas on alternatives or solutions?

**Comment by sharksforarms** [comment]
> 
> 
> I hit this issue while trying to figure out why the strlen_count.py example was not working, I ended up with the same solution (wrapper function). Any ideas on alternatives or solutions?

For those who land on this issue, if you `gdb <program>` and `break strlen` `run` you'll break on some strlen function, mine was `__strlen_avx2`, when I changed the uprobe to this symbol, the example now works.

**Comment by hsane2001** [comment]
Yep, I tried gdb hooking on a memcpy I write and it eneded up getting optimized as __memmove_.... Since there are many implementations I was able to attach uprobe on a regex like this (__memcpy* | __memmove*)

**Comment by sharksforarms** [comment]
@hsane2001 I'd be interested in knowing more what's going on, why doesn't the probe at `strlen` work? What I'm see'ing is that the BPF code gets triggered on a call to strlen (from my program), but the arguments are not what I was expecting.

**Comment by yonghong-song** [comment]
@sharksforarms the following is my interpretation based on my rough memory.
In my libc2.28.so, we have
```
 23748: 0000000000089ac0    53 IFUNC   LOCAL  DEFAULT   14 __GI_strlen
 24299: 000000000015d670   414 FUNC    LOCAL  DEFAULT   14 __strlen_avx2
 24366: 0000000000099d60   412 FUNC    LOCAL  DEFAULT   14 __GI___strlen_sse2
 25278: 0000000000099d60   412 FUNC    LOCAL  DEFAULT   14 __strlen_sse2
 27679: 0000000000089ac0    53 IFUNC   GLOBAL DEFAULT   14 strlen
```
Note that `strlen` is marked as IFUNC (https://jasoncc.github.io/gnu_gcc_glibc/gnu-ifunc.html). In fact, `strlen` is a resolver function. The first time, `strlen` is called, it will try to pick another 'optimal' function based on machine state, say it picks `__strlen_sse2`. Second time you call `strlen`, you directly go to `__strlen_sse2`. So what you really need to trace is `__strlen_sse2`. You can take a look at bcc/examples/tracing/strlen_hist_ifunc.py for an example.
